ml-in-barcelona/server-reason-react
# Handle DomProps with both jsx and html attribute name

## Description

Before, we had the HTML attribute for those `React.JSX.<type>` names, as the JSX isn't needed for SSR.
However, we must have this JSX attribute name for the RSC component.

## How

This PR adds to the DomProps the JSX attribute name data for `React.JSX.Bool` and `React.JSX.String`
Instead of holding only the HTML attribute name, `React.JSX.Bool` and `React.JSX.String` now hold both the HTML attribute name and the JSX attribute.

```diff
  type prop =
-    | Bool of (string * bool)
+    | Bool of ((string * string) * bool)
-    | String of (string * string)
+    | String of ((string * string) * string)
     | Style of string
     | DangerouslyInnerHtml of string
     | Ref of Ref.t
     | Event of string * event
```

### Discussion

@davesnx We talked about it to be a tuple, but could it be a record for a better read? What do you think?
```ocaml
  type prop_name = {
     html: string;
     jsx: string
  }

  type prop =
    | Bool of (prop_name * bool)
    | String of (prop_name * string)
    | Style of string
    | DangerouslyInnerHtml of string
    | Ref of Ref.t
    | Event of string * event
```

## How to test

- Run `make test-watch` into this branch
- Change whatever you want related to this subject, E.g.: 
  - Breaking the code: 
  ```diff
  -  React.JSX.String (("data-user-path", "data-user-path"), "what/the/path");
  +  React.JSX.String (("data-user-path", "data-user-path-error"), "what/the/path");
  ```
- Run `make demo` and see that nothing was changed
